server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -0,0 +1,251 @@
+        this.monitorInterval = monitorInterval;
+        if (scheduled != null && scheduled.cancel()) {
+            scheduleNextRunIfEnabled(new TimeValue(1));
+        }
+    }
+
+    void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        if (scheduled != null) {
+            scheduled.cancel();
+            if (enabled) {
+                scheduleNextRunIfEnabled(new TimeValue(1));
+            }
+        }
+    }
+
+    private void scheduleNextRunIfEnabled(TimeValue time) {
+        if (enabled && threadPool.scheduler().isShutdown() == false) {
+            scheduled = threadPool.schedule(this::monitorHealth, time, ThreadPool.Names.MANAGEMENT);
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // Wait until every node in the cluster is upgraded to 8.4.0 or later
+        if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)) {
+            // Wait until the health metadata is available in the cluster state
+            if (healthMetadataInitialized == false) {
+                healthMetadataInitialized = HealthMetadata.getFromClusterState(event.state()) != null;
+                if (healthMetadataInitialized) {
+                    scheduleNextRunIfEnabled(TimeValue.timeValueMillis(1));
+                }
+            } else if (newHealthNodeSelected(event)) {
+                sendHealth(lastReportedHealth);
+            }
+        }
+    }
+
+    private boolean newHealthNodeSelected(ClusterChangedEvent event) {
+        DiscoveryNode previous = HealthNode.findHealthNode(event.previousState());
+        DiscoveryNode current = HealthNode.findHealthNode(event.state());
+        if (current == null) {
+            return false;
+        } else if (previous == null) {
+            return true;
+        } else {
+            return current.getId().equals(previous.getId()) == false;
+        }
+    }
+
+    // Visible for testing
+    void monitorHealth() {
+        ClusterState clusterState = clusterService.state();
+        HealthMetadata healthMetadata = HealthMetadata.getFromClusterState(clusterState);
+        assert healthMetadata != null : "health metadata should have been initialized.";
+        IndividualNodeHealth previousHealth = this.lastReportedHealth;
+        IndividualNodeHealth currentHealth = new IndividualNodeHealth(diskCheck.getHealth(healthMetadata, clusterState));
+        if (currentHealth.equals(previousHealth) == false) {
+            sendHealth(currentHealth);
+        }
+        scheduleNextRunIfEnabled(monitorInterval);
+    }
+
+    // This method is synchronized to ensure that we keep track of the last reported health state to the health node.
+    // Note: this is for demonstration purposes, when we will be actually sending the information to the health node we might
+    // choose a different way to keep it thread safe.
+    private synchronized void sendHealth(IndividualNodeHealth currentHealth) {
```

Comment:
No synchronisation is needed as the cluster applier thread is .. just one thread (cluster changed is called sequentially) 
I see visibility is ensured by the volatile variables - could you please document the reason why each volatile variable is needed to be volatile? ie. I think for some it's "safe publication of objects written to them" 

Alternatively, should we leave the "send of health" part to another PR to keep this one limited in scope?

Reply:
I will document the volatiles and leave this to the next PR, I still think there is a race but I can elaborate more at the next PR.